tests/functional/test_prometheus_juju_exporter.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/python3
+"""Test exporter daemon."""
+from subprocess import check_call
+
+from helpers import add_machine, get_machines_counts, get_registry_data, remove_machine
+
+
+def test_snap_startup():
+    """Check if the snap is in active state."""
+    assert (
+        check_call(
+            "systemctl is-active --quiet snap.prometheus-juju-exporter.prometheus-juju-exporter.service".split()
+        )
+        == 0  # noqa
+    )
+
+
+def test_connect_registry():
+    """Check if registry can be reached via HTTP requests."""
+    registry_data = get_registry_data()
+    assert registry_data is not None
+
+
+def test_machine_counts():
+    """Test if the snap correct recports total, active, and inactive machine counts."""
+    machine_count = get_machines_counts()
+    registry_data = get_registry_data()
+    machines = [
+        i for i in registry_data.decode().strip().splitlines() if not i.startswith("#")
+    ]
+    assert len(machines) == sum(machine_count.values())
+
+    up_machines = [i for i in machines if i.endswith("1.0")]
+    assert len(up_machines) == machine_count["up"]
+
+    down_machines = [i for i in machines if i.endswith("0.0")]
+    assert len(down_machines) == machine_count["down"]
+
+
+def test_add_machine():
```

Comment:
I think that we should clean up after test function, because like this `test_remove_machine` test depends on this test to be successful.

Reply:
not necessarily. The tests are agnostic of the initial state of the models. They
* Get current status
* Perform changes
* Get status again
* Confirm that the changes are reflected in the new status

I think it's only question about resources. Is it faster to keep deploying new machines, or do we want to wait for removal of machine every time?

Reply:
- If test `test_add_machine` failed, the `test_remove_machine` will also failed due missing machine.
- If test `test_remove_machine` failed and you re-trigger test again, there will be one machine from previous attempt.
- You cannot trigger only `test_remove_machine`.

Reply:
I've updated the function to setup and tear down test model for these two test cases. I think this is the easiest and cleanest way to keep them independent (although the time each test takes is a bit longer).